blockchain/utxocache.go

```diff
@@ -122,3 +129,460 @@ func (ms *mapSlice) deleteMaps() {
+	numMaxElements -= 1
+
+	log.Infof("Pre-alloacting for %d MiB: ", maxTotalMemoryUsage/(1024*1024)+1)
+
+	m := make(map[wire.OutPoint]*UtxoEntry, numMaxElements)
+
+	return &utxoCache{
+		db:                  db,
+		maxTotalMemoryUsage: maxTotalMemoryUsage,
+		cachedEntries: mapSlice{
+			maps:                []map[wire.OutPoint]*UtxoEntry{m},
+			maxEntries:          []int{numMaxElements},
+			maxTotalMemoryUsage: maxTotalMemoryUsage,
+		},
+	}
+}
+
+// totalMemoryUsage returns the total memory usage in bytes of the UTXO cache.
+//
+// This method should be called with the state lock held.
+func (s *utxoCache) totalMemoryUsage() uint64 {
+	// Total memory is the map size + the size that the utxo entries are
+	// taking up.
+	size := uint64(s.cachedEntries.size())
+	size += s.totalEntryMemory
+
+	return size
+}
+
+// fetchEntries returns the UTXO entries for the given outpoints.  The function always
+// returns as many entries as there are outpoints and the returns entries are in the
+// same order as the outpoints.  It returns nil if there is no entry for the outpoint
+// in the UTXO set.
+//
+// This method should be called with the state lock held.
+// The returned entries are NOT safe for concurrent access.
+func (s *utxoCache) fetchEntries(outpoints []wire.OutPoint) ([]*UtxoEntry, error) {
+	entries := make([]*UtxoEntry, len(outpoints))
+	var missingOps []wire.OutPoint
+	var missingOpsIdx []int
+	for i := range outpoints {
+		if entry, ok := s.cachedEntries.get(outpoints[i]); ok {
+			entries[i] = entry
+			continue
+		}
+
+		// At this point, we have missing outpoints.  Allocate them now
+		// so that we never allocate if the cache never misses.
+		if len(missingOps) == 0 {
+			missingOps = make([]wire.OutPoint, 0, len(outpoints))
+			missingOpsIdx = make([]int, 0, len(outpoints))
+		}
+
+		missingOpsIdx = append(missingOpsIdx, i)
+		missingOps = append(missingOps, outpoints[i])
+	}
+
+	// Return early and don't attempt access the database if we don't have any
+	// missing outpoints.
+	if len(missingOps) == 0 {
+		return entries, nil
+	}
+
+	// Fetch the missing outpoints in the cache from the database.
+	dbEntries := make([]*UtxoEntry, len(missingOps))
+	err := s.db.View(func(dbTx database.Tx) error {
+		utxoBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+
+		for i := range missingOps {
+			entry, err := dbFetchUtxoEntry(dbTx, utxoBucket, missingOps[i])
+			if err != nil {
+				return err
+			}
+
+			dbEntries[i] = entry
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Add each of the entries to the UTXO cache and update their memory
+	// usage.
+	//
+	// NOTE: When the fetched entry is nil, it is still added to the cache
+	// as a miss; this prevents future lookups to perform the same database
+	// fetch.
+	for i := range dbEntries {
+		s.totalEntryMemory += dbEntries[i].memoryUsage()
+		s.cachedEntries.put(missingOps[i], dbEntries[i], s.totalEntryMemory)
```

Comment:
Maybe I'm misunderstanding, but why is `s.totalEntryMemory` used as parameter here? Shouldn't it be the size of the single entry?